File: frontend/src/services/knowledgeSync.ts
import { request } from "@/utils/request";

export interface KnowledgeSyncFailure {
  id: number | null;
  path: string;
  stage: string;
  code: string;
  message: string;
  field: string | null;
  line: number | null;
  column: number | null;
}

export interface KnowledgeSyncTask {
  id: number;
  status: "success" | "partial_success" | "failed" | string;
  started_at: string;
  finished_at: string;
  scanned_count: number;
  success_count: number;
  failed_count: number;
  created_count: number;
  updated_count: number;
  restored_count: number;
  unchanged_count: number;
  deleted_count: number;
  failures: KnowledgeSyncFailure[];
}

export interface KnowledgeSyncSchedule {
  enabled: boolean;
  interval_seconds: number;
  running: boolean;
  last_started_at: string | null;
  last_finished_at: string | null;
  last_task_id: number | null;
  last_error: string | null;
}

export interface KnowledgeSyncScheduleUpdate {
  enabled?: boolean;
  interval_seconds?: number;
}

interface KnowledgeSyncTaskListResponse {
  items: KnowledgeSyncTask[];
}

export function triggerKnowledgeSync(): Promise<KnowledgeSyncTask> {
  return request.post<KnowledgeSyncTask>("/knowledge/sync");
}

export async function listKnowledgeSyncTasks(): Promise<KnowledgeSyncTask[]> {
  const response = await request.get<KnowledgeSyncTaskListResponse>("/knowledge/sync/tasks");
  return response.items;
}

export function getKnowledgeSyncTask(taskId: number): Promise<KnowledgeSyncTask> {
  return request.get<KnowledgeSyncTask>(`/knowledge/sync/tasks/${taskId}`);
}

export function getKnowledgeSyncSchedule(): Promise<KnowledgeSyncSchedule> {
  return request.get<KnowledgeSyncSchedule>("/knowledge/sync/schedule");
}

export function updateKnowledgeSyncSchedule(
  payload: KnowledgeSyncScheduleUpdate,
): Promise<KnowledgeSyncSchedule> {
  return request.put<KnowledgeSyncSchedule>("/knowledge/sync/schedule", payload);
}

File: backend/src/techbrain/schemas/knowledge_sync.py
"""Knowledge synchronization API schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class KnowledgeSyncFailureResponse(BaseModel):
    """Synchronization failure response."""

    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    path: str
    stage: str
    code: str
    message: str
    field: str | None = None
    line: int | None = None
    column: int | None = None


class KnowledgeSyncTaskResponse(BaseModel):
    """Synchronization task response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    status: str
    started_at: datetime
    finished_at: datetime
    scanned_count: int
    success_count: int
    failed_count: int
    created_count: int
    updated_count: int
    restored_count: int
    unchanged_count: int
    deleted_count: int
    failures: list[KnowledgeSyncFailureResponse] = []


class KnowledgeSyncTaskListResponse(BaseModel):
    """Synchronization task list response."""

    items: list[KnowledgeSyncTaskResponse]


class KnowledgeSyncScheduleResponse(BaseModel):
    """Periodic synchronization schedule response."""

    model_config = ConfigDict(from_attributes=True)

    enabled: bool
    interval_seconds: int
    running: bool
    last_started_at: datetime | None = None
    last_finished_at: datetime | None = None
    last_task_id: int | None = None
    last_error: str | None = None


class KnowledgeSyncScheduleUpdateRequest(BaseModel):
    """Periodic synchronization schedule update request."""

    enabled: bool | None = None
    interval_seconds: int | None = Field(default=None, ge=60)
